spinalcordtoolbox/math.py

```diff
@@ -7,26 +7,31 @@
 import logging
 import numpy as np
 
-from skimage.morphology import erosion, ball
+from skimage.morphology import erosion, dilation, disk, ball, square, cube
 
 
 logger = logging.getLogger(__name__)
 
 
-def dilate(data, radius):
+def dilate(data, radius, shape, dim=None):
     """
     Dilate data using ball structuring element
-    :param data: 2d or 3d array
-    :param radius: radius of structuring element OR comma-separated int.
-    :return: data dilated
+    :param data: numpy array: 2d or 3d array
+    :param radius: int: Radius of the structuring element. If 0, the convolution has no effect.
+    :param shape: {'square', 'cube', 'disk', 'ball'}
+    :param dim: {0, 1, 2}: Dimension of the array which 2D structural element will be orthogonal to. For example, if
+    you wish to apply a 2D disk kernel in the X-Y plane, leaving Z unaffected, parameters will be: shape=disk, dim=2.
+    :return: numpy array: data dilated
     """
-    from skimage.morphology import dilation, ball
-    if len(radius) == 1:
-        # define structured element as a ball
-        selem = ball(radius[0])
-    else:
-        # define structured element as a box with input dimensions
-        selem = np.ones((radius[0], radius[1], radius[2]), dtype=np.dtype)
+    # TODO: make a build_selem(radius, shape) function called here
+    # TODO: enable custom selem
+    shape = 'ball'
+    # Create structuring element of desired shape and radius
+    # Note: the trick here is to use the variable shape as the skimage.morphology function itself
+    selem = globals()[shape](radius)
```

Comment:
This is clever but it would be safer to just do

```suggestion
    selem = {'square': square, 'cube': cube, 'disk': disk, 'ball': ball}[shape](radius)
```

You could retain some of the cleverness by

```suggestion
    selem = {k: globals()[k] for k in ['square', 'cube', 'disk', 'ball']}[shape](radius)
```

but I am wary of using `globals()` because anything could end up in there and this might allow someone to do some kind of injection attack on this code perhaps a la [prototype pollution](https://snyk.io/vuln/npm:lodash:20180130). But maybe that's an irrelevant concern for scientific computing.

Also, I would split the call from the lookup to make it more readable (I repurpose the variable here but you equally well make up a second name instead)

```suggestion
    shape = {'square': square, 'cube': cube, 'disk': disk, 'ball': ball}[shape]
    selem = shape(radius)
```

Reply:
now using a less elegant (but clearer) switch-case